eZ/Publish/Core/Repository/RoleService.php

```diff
@@ -761,36 +773,31 @@ public function loadRoleByIdentifier($identifier)
             throw new InvalidArgumentValue('identifier', $identifier);
         }
 
-        if ($this->repository->hasAccess('role', 'read') !== true) {
+        $spiRole = $this->userHandler->loadRoleByIdentifier($identifier);
+
+        $role = $this->roleDomainMapper->buildDomainRoleObject($spiRole);
+
+        if (!$this->repository->canUser('role', 'read', $role)) {
             throw new UnauthorizedException('role', 'read');
         }
 
-        $spiRole = $this->userHandler->loadRoleByIdentifier($identifier);
-
-        return $this->roleDomainMapper->buildDomainRoleObject($spiRole);
+        return $role;
     }
 
     /**
-     * Loads all roles.
-     *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to read the roles
+     * Loads all roles, excluding the ones the current user is not allowed to read.
      *
      * @return \eZ\Publish\API\Repository\Values\User\Role[]
      */
     public function loadRoles()
     {
-        if ($this->repository->hasAccess('role', 'read') !== true) {
-            throw new UnauthorizedException('role', 'read');
-        }
+        $roles = array_map(function ($spiRole) {
+            return $this->roleDomainMapper->buildDomainRoleObject($spiRole);
+        }, $this->userHandler->loadRoles());
 
-        $spiRoles = $this->userHandler->loadRoles();
-
-        $roles = [];
-        foreach ($spiRoles as $spiRole) {
-            $roles[] = $this->roleDomainMapper->buildDomainRoleObject($spiRole);
-        }
-
-        return $roles;
+        return array_values(array_filter($roles, function ($role) {
+            return $this->repository->canUser('role', 'read', $role);
```

Comment:
Please perform this check on `$spiRole` object and then build all domain objects.

CS: TBH "inlining" callbacks was never readable to me. See how PHPStorm re-formats that code:
```php
        $roles = array_map(
            function ($spiRole) {
                return $this->roleDomainMapper->buildDomainRoleObject($spiRole);
            },
            $this->userHandler->loadRoles()
        );

        return array_values(
            array_filter(
                $roles,
                function ($role) {
                    return $this->repository->canUser('role', 'read', $role);
                }
            )
        );
```
though the code after my previous remark will be different, this is just an example.